0000-catalog.md

```diff
@@ -0,0 +1,152 @@
+- Feature Name: Catalog
+- Start Date: 2019-07-12
+- RFC PR: 
+- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The
+implementation of *Catalog* will materialize as an assortment of grid products
+that can be shared with one or more organizations, as well as implementaion of
+catalog related functions. The Catalog on grid will contain 6 fields. A
+**catalog_id**, a **catalog_owner** a list of grid **products**, an
+**expiry_date**, an *optional* **Name** for the catalog, and additional
+key:value pairs for custom fields.
+
+In addition any grid product added to a catalog will need to contain a
+***status*** field and can include an optional ***prices*** field.
+
+# Motivation
+[motivation]: #motivation
+
+The Grid Catalog implementation is designed for sharing an assortment of
+products between organizations. Each product in a catalog will contain a number
+of master data elements. Using the Grid Catalog, organizations can share
+collections of related/unrelated products with other participants in their
+network. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Entities
+
+A **catalog** is a collection of
+[products](https://github.com/hyperledger/grid-rfcs/pull/5) that can be shared,
+referenced, or used to do business in a supply chain. A catalog is referenced
+using a **catalog_id**. It has an **owner** (the organization that creates the
+catalog).  It also contains an **expiry_date** as a mechanism for clients to
+exipire the catalog. Also an optional **Name** for the catalog. 
+
+A catalog, like a product, can also have one or more **properties**.  Properties
+are described in the Grid Primitives RFC.  A *property namespace* contains
+multiple *property schemas*.  A property schema associates a name (such as
+“length”) with a data type (such as integer).  
+
+## Transactions
+
+Catalogs are managed by submitting transactions to Hyperledger Grid, which will
+process them with the Grid Catalog smart contract. The following transactions
+are supported:
+
+
+**Catalog actions:**
+* CatalogCreate - create a catalog and store it in state
+* CatalogDelete - removes a catalog from state
+* CatalogReplicate - creates another copy of a catalog that already exists in state
+
+**Catalog operations:**
+* AddProductsToCatalog - adds a list of products specified by their product_id to the catalog 
+* RemoveProductsFromCatalog - removes a product or products from a catalog
+* ActivateProduct - toggles the "status" of a product to an active state
+* DeactivateProduct - toggles the "status" of a product to an inactive state
+* DiscontinueProduct - removes a product from all catalogs it's in
+
```

Comment:
Given that the bulk of these operations are 'about products', or maybe more specifically 'about catalog products', it may make sense to perform the operations on a 'product-like object' instead of a 'catalog-like object'. If there were a location in state referenced by the product identifier that contained a collection of catalog identifiers (and associated 'product catalog attributes'), all the 'catalog operations' listed here could be performed against that structure. This would be separate from the product records, and separate from the catalog record (which would define the catalog identifier and other attributes). This model may make more sense if the cardinality of products to catalogs is materially higher than catalogs to products. Drawbacks are that it makes it difficult to get a list of the products in a catalog without a reporting database. Depending on what we think the extents of 'number of products in a catalog' and/or 'number of catalogs a product is in' are, we may also want to consider other storage models for the collection, including some kind of paginating structure similar to the sensor record storage in track and trace.

Reply:
Also, depending on the sizes expected, storing something like a space-efficient Bloom Filter on either the Catalog or ProductCatalogs records as a maintained existence map may be worthwhile.

Reply:
> Given that the bulk of these operations are 'about products', or maybe more specifically 'about catalog products', it may make sense to perform the operations on a 'product-like object' instead of a 'catalog-like object'. If there were a location in state referenced by the product identifier that contained a collection of catalog identifiers (and associated 'product catalog attributes'), all the 'catalog operations' listed here could be performed against that structure. This would be separate from the product records, and separate from the catalog record (which would define the catalog identifier and other attributes). This model may make more sense if the cardinality of products to catalogs is materially higher than catalogs to products. Drawbacks are that it makes it difficult to get a list of the products in a catalog without a reporting database. Depending on what we think the extents of 'number of products in a catalog' and/or 'number of catalogs a product is in' are, we may also want to consider other storage models for the collection, including some kind of paginating structure similar to the sensor record storage in track and trace.

I've considered an implementation like this, the intention around modeling catalog in its current state was to 'mimic' the state of software catalogs today. I am not strongly opinionated in either approach. Though it would probably warrant a design discussion on this implementation to ensure we're all on the same page. 

Perhaps GS1 would be able to comment on whether the cardinality of products to catalogs is materially higher than catalogs to products. We can do some internal research ourselves

Reply:
I think I need to understand more about the overall approach.  Today,  there typically is 1 supplier catalog (as it is today), all products listed, with certain products shared with specific trading partners? So - generally more products than catalogs. Numbers of products are exponentially higher than catalogs. HOWEVER - are you guys thinking there is a 1:1 catalog for every trading partner and the products shared between those parties constitutes the catalog (NOT how it works today)

Reply:
Good to note that suppliers have one "master" catalog and share specific products with trading partners. Perhaps an introduction of a catalog_type enum would make sense. A "MASTER_CATALOG" could be the all-encompassing catalog of products, while a "TRADING_PARTNER_CATALOG" could be a subset of those products with specific prices for the trading partner.